core/runtime/src/main/java/io/quarkus/runtime/configuration/QuarkusConfigFactory.java

```diff
@@ -19,21 +19,37 @@ public QuarkusConfigFactory() {
         // todo: replace with {@code provider()} post-Java 11
     }
 
+    @Override
     public SmallRyeConfig getConfigFor(final SmallRyeConfigProviderResolver configProviderResolver,
             final ClassLoader classLoader) {
         if (config == null) {
-            return ConfigUtils.configBuilder(true, true, LaunchMode.NORMAL).build();
+            // Remember the config so that we can uninstall it when "setConfig" is next called
+            config = ConfigUtils.configBuilder(true, true, LaunchMode.NORMAL).build();
         }
         return config;
     }
 
     public static void setConfig(SmallRyeConfig config) {
         SmallRyeConfigProviderResolver configProviderResolver = (SmallRyeConfigProviderResolver) SmallRyeConfigProviderResolver
                 .instance();
-        configProviderResolver.releaseConfig(Thread.currentThread().getContextClassLoader());
-        QuarkusConfigFactory.config = config;
+        // Uninstall previous config
         if (QuarkusConfigFactory.config != null) {
+            configProviderResolver.releaseConfig(QuarkusConfigFactory.config);
```

Comment:
While unlikely, it is possible to release / register the Config outside this Factory, invalidating the cached referenced when calling `getConfigFor` for `SmallRyeConfigProviderResolver`

It is safer to just release whatever Config is registered to the TCCL or a specific CL (that is why we added that the release method per CL).

Is there a reason for going back to release per Config and not per CL?

Reply:
> While unlikely, it is possible to release / register the Config outside this Factory, invalidating the cached referenced when calling `getConfigFor` for `SmallRyeConfigProviderResolver`

If someone calls `SmallRyeConfigProviderResolver` release/register, I'm going to guess that invalidating the cached reference in `SmallRyeConfigProviderResolver` is precisely what they want, so I don't see the problem?

> Is there a reason for going back to release per Config and not per CL?

Yes.

`QuarkusConfigFactory#getConfigFor` actually provides a default config for whenever `SmallRyeConfigProviderResolver` is asked for the config of a classloader and none is registered yet. As a result, there are many more classloaders that can end up bound to `QuarkusConfigFactory.config` than just the TCCL when we call `setConfig`.

How it works: when someone retrieves the config via `SmallRyeConfigProviderResolver` for a given classloader, and that classloader is the not the TCCL that was active when calling `setConfig`,  `SmallRyeConfigProviderResolver` will call `QuarkusConfigFactory.getConfigFor`, and ultimately insert an entry in `SmallRyeConfigProviderResolver#configsForClassLoader` with that classloader as key and the config as value .

If we only release config for the TCCL, this entry for another classloader will effectively persist forever, and so will the reference to the classloader, and you got yourself a metaspace leak.
That's why I call `releaseConfig(config)`: to invalidate the cache of *all* classloaders that used this config, and avoid leaks.

Note that we *also* release per-CL where it makes sense: see `releaseTCCLConfig` and how it's called in `ConfigRecorder`. I changed the way we release because the existing calls to `setConfig(null)` were actually using a `QuarkusConfigFactory` and a `SmallRyeConfigProviderResolver` from a different classloader than the initial call to `setConfig(<non-null>)` in the generated config class.

Anyway... it's way too complex for its own good, and that's why I wanted to make Smallrye Config parent-first, but that didn't fly because people had performance concerns. I tried many alternatives before settling for this one, and most were just breaking the tests. This one fixes the leaks **and** doesn't break the tests.

Reply:
Alright, fair point, I did not know that `registerConfig` failed when there was already a registered config. Thanks for pointing that out.

I amended the commit to add a `releaseConfig(TCCL)` as you suggested. Is this better?

Reply:
Yes, thank you :)